## <center>Car Accident Severity: Data</center>

In this section, we will be importing the dataset, exploring it and understanding it to select relevant data with the appropriate features as per the needs and wants of the problem.

Firstly, importing some important python libraries:

In [56]:
import pandas as pd
import numpy as np
import matplotlib as mlp
import matplotlib.pyplot as plt


In [3]:
!conda install -c conda-forge folium=0.5.0 --yes

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - folium=0.5.0


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    ca-certificates-2020.6.20  |       hecda079_0         145 KB  conda-forge
    vincent-0.4.4              |             py_1          28 KB  conda-forge
    folium-0.5.0               |             py_0          45 KB  conda-forge
    altair-4.1.0               |             py_1         614 KB  conda-forge
    openssl-1.1.1h             |       h516909a_0         2.1 MB  conda-forge
    certifi-2020.6.20          |   py36h9f0ad1d_0         151 KB  conda-forge
    branca-0.4.1               |             py_0          26 KB  conda-forge
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    ------------------------------------------------------------
                       

Now importing the data set from the remote source are taking a perfunctory look at it:

In [57]:
df= pd.read_csv("http://data-seattlecitygis.opendata.arcgis.com/datasets/5b5c745e0f1f48e7a53acec63a0022ab_0.csv?outSR=%7B%22latestWkid%22%3A2926%2C%22wkid%22%3A2926%7D")
df.head()

X              Y  OBJECTID  INCKEY  COLDETKEY REPORTNO  \
0  1.268354e+06  265256.609668         1    1003       1003  3503158   
1  1.259316e+06  202173.388163         2   56200      56200  1795087   
2  1.264342e+06  259613.000382         3  327037     328537  E979380   
3  1.279221e+06  222017.872023         4  327278     328778  E996362   
4  1.262205e+06  242179.124204         5    1248       1248  3645424   

      STATUS      ADDRTYPE   INTKEY  \
0    Matched         Block      NaN   
1    Matched         Block      NaN   
2    Matched  Intersection  37122.0   
3  Unmatched  Intersection  30602.0   
4  Unmatched         Block      NaN   

                                           LOCATION  ... ROADCOND  \
0    AURORA AVE N BETWEEN N 117TH PL AND N 125TH ST  ...      Dry   
1  35TH AVE SW BETWEEN SW MORGAN ST AND SW HOLLY ST  ...      Dry   
2                        3RD AVE NW AND NW 100TH ST  ...      Wet   
3                M L KING JR WAY S AND S JACKSON ST  ...      NaN   
4       W EWING ST BETWEEN 6TH AVE W AND W EWING PL  ...      NaN   

                 LIGHTCOND PEDROWNOTGRNT SDOTCOLNUM SPEEDING  ST_COLCODE  \
0                 Daylight             Y        NaN      NaN          45   
1  Dark - Street Lights On           NaN  6015003.0      NaN           0   
2                 Daylight           NaN        NaN      NaN          10   
3                      NaN           NaN        NaN      NaN               
4                      NaN           NaN        NaN      NaN               

                               ST_COLDESC  SEGLANEKEY  CROSSWALKKEY  \
0                  Vehicle - Pedalcyclist       15057             0   
1  Vehicle going straight hits pedestrian           0             0   
2                       Entering at angle           0             0   
3                                     NaN           0             0   
4                                     NaN           0             0   

   HITPARKEDCAR  
0             N  
1             N  
2             N  
3             N  
4             N  

[5 rows x 40 columns]

In [59]:
df.size

8861000

In [60]:
df.shape

(221525, 40)

In [61]:
df.columns

Index(['X', 'Y', 'OBJECTID', 'INCKEY', 'COLDETKEY', 'REPORTNO', 'STATUS',
       'ADDRTYPE', 'INTKEY', 'LOCATION', 'EXCEPTRSNCODE', 'EXCEPTRSNDESC',
       'SEVERITYCODE', 'SEVERITYDESC', 'COLLISIONTYPE', 'PERSONCOUNT',
       'PEDCOUNT', 'PEDCYLCOUNT', 'VEHCOUNT', 'INJURIES', 'SERIOUSINJURIES',
       'FATALITIES', 'INCDATE', 'INCDTTM', 'JUNCTIONTYPE', 'SDOT_COLCODE',
       'SDOT_COLDESC', 'INATTENTIONIND', 'UNDERINFL', 'WEATHER', 'ROADCOND',
       'LIGHTCOND', 'PEDROWNOTGRNT', 'SDOTCOLNUM', 'SPEEDING', 'ST_COLCODE',
       'ST_COLDESC', 'SEGLANEKEY', 'CROSSWALKKEY', 'HITPARKEDCAR'],
      dtype='object')

So there are 40 columns with different names. The description of these names can be found <a href= https://s3.us.cloud-object-storage.appdomain.cloud/cf-courses-data/CognitiveClass/DP0701EN/version-2/Metadata.pdf>here</a>. A brief summary of the dataframe along with the datatype of the columns can be found below.

Let us rename these column names to make it more convenient for 

In [62]:
df.rename(columns={'SEVERITYCODE': 'severity_code', 'X':'longitude', 'Y': 'latitude',
                   'ADDRTYPE':'addr_type', 'LOCATION': 'location','SEVERITYDESC':'severity_desc', 'COLLISIONTYPE':'collision_type',
                   'PERSONCOUNT':'person_count', 'PEDCOUNT': 'ped_count', 'PEDCYLCOUNT': 'ped_cycle_count', 'VEHCOUNT': 'veh_count',
                   'INCDTTM': 'incident_date&time', 'JUNCTIONTYPE': 'junc_type', 'SDOT_COLCODE': 'case_code', 'SDOT_COLDESC': 'case_desc',
                   'UNDERINFL':'under_infl', 'WEATHER': 'weather', 'ROADCOND': 'roadcond', 'LIGHTCOND': 'light_cond',
                   'ST_COLCODE': 'st_code', 'ST_COLDESC': 'st_desc', 'HITPARKEDCAR':'hit_parked_car', 'SPEEDING':'speeding', 
                   'FATALITIES':'fatalities', 'INJURIES':'injuries', 'SERIOUSINJURIES':'serious_injuries', 'INCDATE':'incident_date', 
                   'INATTENTIONIND':'inattention_involved', 'PEDROWNOTGRNT':'ped_row_not_granted', 'SDOTCOLNUM':'sdot_collision_num'}, inplace=True)

In [63]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 221525 entries, 0 to 221524
Data columns (total 40 columns):
longitude               214050 non-null float64
latitude                214050 non-null float64
OBJECTID                221525 non-null int64
INCKEY                  221525 non-null int64
COLDETKEY               221525 non-null int64
REPORTNO                221525 non-null object
STATUS                  221525 non-null object
addr_type               217813 non-null object
INTKEY                  71936 non-null float64
location                216935 non-null object
EXCEPTRSNCODE           101122 non-null object
EXCEPTRSNDESC           11779 non-null object
severity_code           221524 non-null object
severity_desc           221525 non-null object
collision_type          195212 non-null object
person_count            221525 non-null int64
ped_count               221525 non-null int64
ped_cycle_count         221525 non-null int64
veh_count               221525 non-null int64
in

Now that we have acquired the data, let us try to comprehend it to get some idea on which attributes or features should be selected for the modelling.

### Data Understanding

The aim of this work is to predict the severity of a car accident based on various attributes. The columns 'severity_code and 'severity_desc' depict the code that corresponds to the level of severity as classified by SDOT and the description of the severity of the collision respectively. Let us have a look at the data distribution according to the severity of the incident.

**Collision Severity**

In [64]:
df.groupby(['severity_code', 'severity_desc']).size().to_frame('counts')

counts
severity_code severity_desc                         
0             Unknown                          21615
1             Property Damage Only Collision  137671
2             Injury Collision                 58783
2b            Serious Injury Collision          3105
3             Fatality Collision                 350

We can see that majority of collisions are non-fatal in nature, with most collisions only leading to property damage an injuries. Now, let us have a look at the various collision types recorded in the dataset.

**Collision Type**

In [65]:
df['collision_type'].value_counts().to_frame('Counts').sort_values(by='Counts',ascending=False)

Counts
Parked Car   48551
Angles       35573
Rear Ended   34691
Other        24588
Sideswipe    18891
Left Turn    14115
Pedestrian    7666
Cycles        5932
Right Turn    3017
Head On       2188

We see that collision with parked cars was the most frequent collision type while head on Collisions were the least common. These collisions happened at different locations as shown in the following section.

**Collision Address type and Junction Type**

In [66]:
df.groupby(['addr_type', 'junc_type']).size().to_frame('Counts')

Counts
addr_type    junc_type                                                
Alley        At Intersection (but not related to intersection)       1
             Driveway Junction                                      66
             Mid-Block (not related to intersection)               195
Block        At Intersection (but not related to intersection)       1
             At Intersection (intersection related)                  4
             Driveway Junction                                   11426
             Mid-Block (but intersection related)                24388
             Mid-Block (not related to intersection)            100976
             Ramp Junction                                         153
             Unknown                                                12
Intersection At Intersection (but not related to intersection)    2491
             At Intersection (intersection related)              69189
             Mid-Block (but intersection related)                    1
             Mid-Block (not related to intersection)                15
             Ramp Junction                                          35
             Unknown                                                 6

We see that most collisions occured at the Blocks and Intersections. This may be due to the fact that these locations are the points where traffic merges and therefore are much prone to conflicts. Now let us see the description of the weather conditions during the time of the collision. 

**Weather** 

In [67]:
df['weather'].value_counts().to_frame('Counts')

Counts
Clear                     114738
Raining                    34036
Overcast                   28552
Unknown                    15131
Snowing                      919
Other                        860
Fog/Smog/Smoke               577
Sleet/Hail/Freezing Rain     116
Blowing Sand/Dirt             56
Severe Crosswind              26
Partly Cloudy                 10
Blowing Snow                   1

We find that most collisions occur during clear weather condition followed by raining and overcast conditions. This might be because the drivers tend to avoid driving when the weather condition is worse. Next, let us see the road conditons at the spot of collisions.

**Road Condition**

In [37]:
df['roadcond'].value_counts().to_frame('Counts')

Counts
Dry             128588
Wet              48734
Unknown          15139
Ice               1232
Snow/Slush        1014
Other              136
Standing Water     119
Sand/Mud/Dirt       77
Oil                 64

Most collisions seem to have happened on dry and wet roads. 

**Light Condition**

We can see the light conditions during the collision in the table below. 

In [68]:
df['light_cond'].value_counts().to_frame('counts')

counts
Daylight                  119492
Dark - Street Lights On    50133
Unknown                    13532
Dusk                        6082
Dawn                        2609
Dark - No Street Lights     1579
Dark - Street Lights Off    1239
Other                        244
Dark - Unknown Lighting       23

So most collisions occurred during good lighting conditions during daytime and at night with street lights on. Another factor that can cause accidents is the condition of the road user being intoxicated i.e. under the influence of alcohol and drugs. The next section shows the distibution of whether the driver involved was under the influence of drugs or alcohol.  

**Under the Influence of alcohol/drugs**

In [39]:
df['under_infl'].value_counts().to_frame('counts')

counts
N  103927
0   81676
Y    5399
1    4230

We see that this column has different entries used, which are just the duplicate of themselves .Hence changing them to a unique numeric data which will be good for processing the data. 

In [69]:
df['under_infl'].replace(to_replace={'Y':1, 'N':0, '1':1, '0':0}, inplace=True)
df['under_infl'].value_counts().to_frame('counts')

counts
0.0  185603
1.0    9629

So we can see that in most cases people were not intoxicated. Now let us have a look at the distribution of various road users involved in these collisions.

**Collisions by Road User** 

In [70]:
df[['person_count', 'ped_count', 'ped_cycle_count', 'veh_count']].describe()

person_count      ped_count  ped_cycle_count      veh_count
count  221525.000000  221525.000000    221525.000000  221525.000000
mean        2.226941       0.038118         0.027360       1.730482
std         1.470050       0.201766         0.164537       0.829754
min         0.000000       0.000000         0.000000       0.000000
25%         2.000000       0.000000         0.000000       2.000000
50%         2.000000       0.000000         0.000000       2.000000
75%         3.000000       0.000000         0.000000       2.000000
max        93.000000       6.000000         2.000000      15.000000

We see that mostly vehicle owners were involved in these accidents, with an average of 2.23 people and 1.73 vehicle involved in each collision. Now let us move to the Data Cleaning and Preprocessing stage.


### Data Cleaning & Preprocessing

The first step in this would be to take care of the missing or Nan values in the dataframe. If Nan values are present, we would either delete the entire row/column or replace the null value to some other suitable value, depending on the requirement. Let us find ut how many null values are present in each column.

In [71]:
df.isnull().sum()

longitude                 7475
latitude                  7475
OBJECTID                     0
INCKEY                       0
COLDETKEY                    0
REPORTNO                     0
STATUS                       0
addr_type                 3712
INTKEY                  149589
location                  4590
EXCEPTRSNCODE           120403
EXCEPTRSNDESC           209746
severity_code                1
severity_desc                0
collision_type           26313
person_count                 0
ped_count                    0
ped_cycle_count              0
veh_count                    0
injuries                     0
serious_injuries             0
fatalities                   0
incident_date                0
incident_date&time           0
junc_type                11974
case_code                    1
case_desc                    1
inattention_involved    191337
under_infl               26293
weather                  26503
roadcond                 26422
light_cond               26592
ped_row_

STATUS, INTKEY,OBJECTID, INCKEY, COLDETKEY, REPORTNO, EXCEPTRSNCODE, EXCEPTRSNDESC, incident_date, inattention_involved, ped_row_not_granted, sdot_collision_num, SEGLANEKEY, CROSSWALKKEY, addr_type, location, st_code, st_desc, case_code, case_desc, incident_date&time has a lot of missing values and they are not useful and irrelevant for this dataset. So they are dropped.

In [72]:
df.drop(['STATUS', 'INTKEY', 'OBJECTID', 'INCKEY', 'COLDETKEY', 'REPORTNO', 'EXCEPTRSNCODE', 'EXCEPTRSNDESC', 'incident_date', 'inattention_involved', 'ped_row_not_granted', 'sdot_collision_num', 'SEGLANEKEY', 'CROSSWALKKEY', 'addr_type', 'location', 'st_code', 'st_desc', 'case_code', 'case_desc', 'incident_date&time'], axis=1, inplace=True)

In [73]:
df.shape

(221525, 19)

In [74]:
df.isnull().sum()

longitude             7475
latitude              7475
severity_code            1
severity_desc            0
collision_type       26313
person_count             0
ped_count                0
ped_cycle_count          0
veh_count                0
injuries                 0
serious_injuries         0
fatalities               0
junc_type            11974
under_infl           26293
weather              26503
roadcond             26422
light_cond           26592
speeding            211596
hit_parked_car           0
dtype: int64

The speeding column has a lot of missing values and consists of only 'y' in cases that involved speeding. So, converting Y->1 and nan->0 to make the speeding column consistent and get rid of all Nan values.

In [80]:
df['speeding'].replace(to_replace={'Y':1, np.nan:0, '1':1, '0':0}, inplace=True)
df['speeding'].value_counts().to_frame('counts')

counts
0  211596
1    9929

The dataset has some columns with useless data like unknown and others which won't fall under any of the category and they are a kind of outliers which ruin the dataset .Hence they are all converted to nan, thereby, paving way to group them under the category of missing values.

In [81]:
df.replace(to_replace={'Unknown':np.nan, 'Other':np.nan}, inplace=True)

Now dropping all rows with missing data in the dataset.

In [82]:
df.dropna(axis=0, inplace=True)

In [83]:
df.shape

(148112, 19)

In [84]:
df.isnull().sum()

longitude           0
latitude            0
severity_code       0
severity_desc       0
collision_type      0
person_count        0
ped_count           0
ped_cycle_count     0
veh_count           0
injuries            0
serious_injuries    0
fatalities          0
junc_type           0
under_infl          0
weather             0
roadcond            0
light_cond          0
speeding            0
hit_parked_car      0
dtype: int64

So, as we can see, we are only left with 19 columns in our dataset, with each column being devoid of missing values. This concludes the stage of data cleaning. Next we need to carry out some data visualization as a part of exploratory data analysis before we move on to Modelling.